### PR TITLE
Add Tor hybrid connectivity mode

### PR DESCRIPTION
**Pull request by adriansmares**
###
Closes https://github.com/lightningnetwork/lnd/issues/5042

This PR allows the Tor dialer (`ProxyNet`) to avoid routing traffic via Tor for targets which can be contacted via clearnet. 

#### Pull Request Checklist

- [x] If this is your first time contributing, we recommend you read the [Code
  Contribution Guidelines](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md)
- [x] All changes are Go version 1.12 compliant
- [x] The code being submitted is commented according to [Code Documentation and Commenting](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#CodeDocumentation)
- [ ] For new code: Code is accompanied by tests which exercise both
  the positive and negative (error paths) conditions (if applicable)
- [ ] For bug fixes: Code is accompanied by new tests which trigger
  the bug being fixed to prevent regressions
- [x] Any new logging statements use an appropriate subsystem and
  logging level
- [x] Code has been formatted with `go fmt`
- [ ] Protobuf files (`lnrpc/**/*.proto`) have been formatted with
  `make rpc-format` and compiled with `make rpc`
- [x] New configuration flags have been added to `sample-lnd.conf`
- [x] For code and documentation: lines are wrapped at 80 characters
  (the tab character should be counted as 8 characters, not 4, as some IDEs do
  per default)
- [ ] Running `make check` does not fail any tests
- [x] Running `go vet` does not report any issues
- [ ] Running `make lint` does not report any **new** issues that did not
  already exist
- [ ] All commits build properly and pass tests. Only in exceptional
  cases it can be justifiable to violate this condition. In that case, the
  reason should be stated in the commit message.
- [x] Commits have a logical structure according to [Ideal Git Commit Structure](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#IdealGitCommitStructure)

### CODE REVIEW

File: sample-lnd.conf [hunk 1]
```diff
@@ -834,6 +834,13 @@ litecoin.node=ltcd
 ; Allow outbound and inbound connections to be routed through Tor
 ; tor.active=true
 
+; Allow the node to connect to non-onion services directly via clearnet. This
+; allows the node operator to use direct connections to peers not running behind
+; Tor, thus allowing lower latency and better connection stability.
+; WARNING: This option will reveal the source IP address of the node, and should
+; be used only if privacy is not a concern.
+; tor.directconnections=true
```
Comment (Roasbeef) [inline sample-lnd.conf:842]: Solid helper text here, I can't help but think maybe we should modify the _name_ of this new option though... 🤔

Don't have any solid suggestions here, but will circle back if I can think of something.

**Review by Roasbeef — COMMENTED** [review]
Thanks for picking this up! The diff turned out to be much smaller than I expected which is great. 

Just completed an initial review pass, and didn't see anything glaring jump out. Will put this on my queue of PRs to test out to verify things work as advertised.

**Comment by gkrizek** [comment]
Not immediately clear to me from the PR if this affects Neutrino or not. Would Neutrino us this same method? For example if I started LND with `--tor.active --tor.v3 --tor.directconnections --neutrino.connect=1.1.1.1`. would Neutrino connect to `1.1.1.1` on Clearnet or through Tor?

File: sample-lnd.conf (hunk 1 above)
Comment (Kixunil) [reply sample-lnd.conf:842]: I would rename every occurrence of "direct" with "clearnet". (except for "directly via clearnet") So the option would be `tor.clearnetconnections`

File: tor/tor.go [hunk 2]
```diff
@@ -111,9 +118,22 @@ func dial(address, socksAddr string, streamIsolation bool,
...
 
+	clearDialer := &net.Dialer{Timeout: timeout}
+	if directConnections {
+		host, _, err := net.SplitHostPort(address)
+		if err != nil {
+			return nil, err
+		}
+
+		// The SOCKS proxy is skipped if the target
+		// address is not an union address.
```
Comment (Kixunil) [inline tor/tor.go:129]: "is not an onion address"

**Review by Kixunil — COMMENTED** [review]
Awesome progress, thanks a lot!

**Comment by C-Otto** [comment]
Running on c-otto.de (027ce055380348d7812d2ae7745701c9f93e70c1adeb2657f053f91df4f2843c71@157.90.112.145:9735). According to tcpdump I still send/receive clearnet data from my peer bfx-lnd1, and I'm able to connect to yalls-tor (which only has an onion address).

Thanks a lot!

**Comment by adriansmares** [comment]
> 
> 
> Not immediately clear to me from the PR if this affects Neutrino or not. Would Neutrino us this same method? For example if I started LND with `--tor.active --tor.v3 --tor.directconnections --neutrino.connect=1.1.1.1`. would Neutrino connect to `1.1.1.1` on Clearnet or through Tor?

https://github.com/lightningnetwork/lnd/blob/b04efec13066362529e914e1805880b3d20fad7d/lnd.go#L1777-L1831

It would go via clearnet. The Neutrino dialer and name resolver use the configured network (`ClearNet`/`ProxyNet`). Since you've enabled Tor, `cfg.net` is a `ProxyNet`. The direct connection setting makes `ProxyNet` call clearnet targets directly, instead of using the SOCKS proxy provided by Tor.

**Comment by Kixunil** [comment]
> I would find it weird that I can still connect to clearnet peers if I have `tor.clearnetconnections=false` - that name makes sense if you know what we're trying to name, but that's not the point of a name.

Ah, true, it may look like it disables non-`.onion` connections. Maybe `tor.nontorconnections`?

**Comment by yuvadm** [comment]
Testing this build now and it seems to be working properly. Up until now my node was cleartext only, and now I'm testing `--tor.active` and `--tor.directconnections`.

My node still seems to publish a clearnet external IP, and yet I can resolve `.onion` peers.

Using `lsof` to verify the connections, I see `lnd`:
- listening on `*:9735`
- connecting to SOCKS on `localhost:9050`
- connecting directly to clearnet peers on `1.2.3.4:9735`
- connecting to various tor nodes on random ports

AFAICT all is working properly.

**Comment by adriansmares** [comment]
> 
> 
> > I would find it weird that I can still connect to clearnet peers if I have `tor.clearnetconnections=false` - that name makes sense if you know what we're trying to name, but that's not the point of a name.
> 
> Ah, true, it may look like it disables non-`.onion` connections. Maybe `tor.nontorconnections`?

I think it suffers from the same problem. I think the helper text should be the driving guidance here, not the name only.

**Comment by Kixunil** [comment]
As a big fan of naming things clearly it's hard for me to give up. :) `mixedconnectiontypes`? Also Tor != onion but few people understand the distinction.

**Comment by adriansmares** [comment]
> 
> 
> As a big fan of naming things clearly it's hard for me to give up. :) `mixedconnectiontypes`? Also Tor != onion but few people understand the distinction.

`skip-proxy-for-clearnet-targets` then ? At heart, that's what the hybrid connectivity means here - that you skip the Tor proxy for clearnet targets.

**Comment by Kixunil** [comment]
Sure, `skip-proxy-for-clearnet-targets` sounds very good. Maybe `skip-tor-proxy-for-clearnet-targets` would be better but it's starting to be quite long.

**Comment by lschmierer** [comment]
Running this for a few days now. Seems to work fine so far! 👍

File: lncfg/tor.go [hunk 3]
```diff
@@ -6,6 +6,7 @@ type Tor struct {
 	SOCKS             string `long:"socks" description:"The host:port that Tor's exposed SOCKS5 proxy is listening on"`
 	DNS               string `long:"dns" description:"The DNS server as host:port that Tor will use for SRV queries - NOTE must have TCP resolution enabled"`
 	StreamIsolation   bool   `long:"streamisolation" description:"Enable Tor stream isolation by randomizing user credentials for each connection."`
+	DirectConnections bool   `long:"directconnections" description:"Allow the node to establish direct connections to services not running behind Tor."`
```
Comment (wpaulino) [inline lncfg/tor.go:9]: From all of the suggestions in the PR discussion, I like `skip-proxy-for-clearnet-targets` the most.

**Review by wpaulino — APPROVED** [review]

**Comment by Roasbeef** [comment]
Had to kick Travis manually to trigger an itest run (as you're a new contributor, part of their like anit-spam rules).

**Review by Roasbeef — APPROVED** [review]
LGTM 🥽

Excellent first time contribution! 

Only thing this is missing now is an entry in the release notes for the upcoming 0.14 release: https://github.com/lightningnetwork/lnd/blob/master/docs/release-notes/release-notes-0.14.0.md

**Comment by 3nprob** [comment]
This is a nice improvement in isolation, but I have one concern with it: It doesn't address scenarios like:

* Skip tor only for `127.0.0.0/8`
* Skip tor only for `127.0.0.0/8` and `*.internal.mycorp.com`

(see #5647 for example)

Merging it as-is will kick those down the road associated with either new overloaded parameters, or introduce breaking changes when it gets replaced later on.

A proposal: Change the configuration parameter from `bool` to `[]string`, with a list of subnets and hostnames to skip the proxy for. I'm pretty much talking about making the new parameter an equivalent to the conventional [`NO_PROXY`](https://about.gitlab.com/blog/2021/01/27/we-need-to-talk-no-proxy/) env var.

**Comment by 3nprob** [comment]
@Roasbeef @adriansmares I made a PR on @adriansmares repo (couldnt seem to do it in this repo with changes based off this unil it's merged)

Let me know what you think!

**Comment by Roasbeef** [comment]
Merging this as is, since it's been complete for some time. 

@3nprob you can make a new PR proposing your desired changes, but see my comment on the other issue.

**Comment by 3nprob** [comment]
New PR: https://github.com/lightningnetwork/lnd/pull/5651

**Pull request merged**